doc/reference/trd-shared-allow-syscalls.md

```diff
@@ -0,0 +1,95 @@
+reading restriction is because the contents of the buffer may be in an
+intermediate state and so not consistent with expected data
+models. Ensuring every system call driver maintains consistency in the
+presence of arbitrary userspace reads is too great a programming
+burden for an unintended use case.
+
+However, there can be cases when it is necessary for userspace to be
+able to read a buffer without first revoking it from the kernel with a
+Read-Write Allow. These cases are situations when the cost of a
+Read-Write Allow system call is an unacceptable overhead for
+accessing the data.
+
+This allows capsules that support the shared allow call to communicate with
+applications without buffers needing to be re-allowed. For example a capsule
+might want to report statistics to a userspace app. It could do this by letting
+the app perform a shared allow call to allocate a buffer. Then the capsule can
+write statistics to the buffer and at any time the app can read the statistics
+from the buffer.
+
+2 System Call API
+=================================
+
+2.1 Shared Allow (Class ID: 7)
+---------------------------------
+The shared allow syscall follows the same expectations and requirements
+as described for the Read-Write syscall in section TRD104 4.4, with the
+exceptionthat apps are explicity allowed to read/write buffers that have
+been passed to the kernel.
+
+As simultaneous accesses to a buffer from both userspace and the kernel can
+cause issues if not implemented properly. For example a userspace app could
+read partially written data or the kernel could act on partially written data.
+This would result in obscure timing bugs that are hard to detect. Due to this
+each capsule using the Shared Allow mechanism MUST document, in a Draft or
+Final Documentary TRD, how consistency of the data exchanged through this
+shared memory region is achieved.
+
+Examples for mechanisms to ensure consistent data reads/writes can be driver state
+machines, precisely specifying when writes will be issued to the buffer by the
+kernel, or monotonically increasing counters along the data for userspace to ensure
+a read operation yielded consistent data.
+
+For an example of a monotomic counter implementation look at the
+`doc/syscalls/10001_ros.md` description.
+
+3 libtock-c Userspace Library Methods
+=====================================
+
+3.1 Shared Allow
+---------------------------------
+
+The read-write allow system call has this function prototype:
+
+```c
+typedef struct {
+  bool success;
+  void* ptr;
+  size_t size;
+  tock_error_t error;
+} allow_shared_return_t;
+
+allow_shared_return_t allow_shared(uint32_t driver, uint32_t allow, void* ptr, size_t size);
+```
+
+The `success` field indicates whether the call succeeded.
+If it failed, the error code is stored in `error`. If it succeeded,
+the value in `error` is undefined. `ptr` and `size` contain the pointer
+and size of the passed buffer.
```

Comment:
This should also explain what the `allow` parameter is [even if it's just a duplication of some of the 104 §4.4 text, locality of reference is valuable here]